### PR TITLE
generate flyway migrations from hibernate update scripts

### LINKED ISSUE #43723: Generate Liquibase Database Migration from hibernate model
### Description

Today, there is no real way to generate db migrations.

First I need to do the modifications in my entity, then I lookup the create script in the dev UI and I create a new migration manually.
This is quite redundant since my entities already contains all the info to create the migration (at least draft a first version of it).

I remember when I first started coding in PHP, doctrine was doing this very well: https://symfony.com/bundles/DoctrineMigrationsBundle/current/index.html

### Implementation ideas

Liquidbase provide a diff command that compute the difference between two database, and a generate changelog command that generate migrations from this diff.

Instead of comparing real databases, we could compare a virtual database created from hibernate persistence model and the local dev db.

There is already some work around a hibernate virtual db, but it is not working with quarkus:
https://github.com/liquibase/liquibase-hibernate

This could then be executed either from the dev UI, or by registering a new maven / gradle task.

### PR DESCRIPTION
**Pull request by rmanibus**
fix #43723 

Current Process: When creating a database migration, the workflow is quite cumbersome:

- Modify the entity.
- Check the update script in the dev UI (when accessible—sometimes the app doesn't even start, making it unavailable).
- Manually create a new migration file.
- Customize the update script as needed.

This process involves redundancy, as the entity already contains the necessary information to at least generate a draft migration.

New Workflow: With this pull request, the process is simplified to:

- Modify the entity.
- Customize the update script as needed.

Migration naming convention:

- major version is extracted from the previous migration
- minor version is the current timestamp (current format is yyyyMMddHHmmss, but I think flyway might support something more readable like yyyy_MM_dd_HH_mm_ss)

### CODE REVIEW

File: extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeUpdateFromHibernateTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    @Test
+    public void testGenerateMigrationFromHibernate() throws Exception {
+
+        Map<String, Object> params = Map.of("ds", "<default>");
+        JsonNode devuiresponse = super.executeJsonRPCMethod("update", params);
+
+        Assertions.assertNotNull(devuiresponse);
+        String type = devuiresponse.get("type").asText();
+        Assertions.assertNotNull(type);
+        Assertions.assertEquals("success", type);
```
Comment (rmanibus) [inline extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeUpdateFromHibernateTest.java:39]: how can I actually check the content if the generated file here ?

Comment (gsmet) [reply extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeUpdateFromHibernateTest.java:39]: @phillip-kruger I think we need your help here :)

**Comment by gsmet** [comment]
Thanks for the contribution. Could you clarify how it works and the cases that are covered?

**Comment by rmanibus** [comment]
> Thanks for the contribution. Could you clarify how it works and the cases that are covered?

@gsmet just updated the PR description

**Comment by rmanibus** [comment]
@gsmet I have been able to complete the test, this is now ready for review

**Review by yrodiere — COMMENTED** [review]
Looks like we forgot about this one, sorry...

@gsmet it looks like this is a valuable feature and it at least goes in the right direction, so I'd say let's merge it?

Liquibase can get a similar feature later, when someone has time/motivation to introduce it.

File: extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js [hunk 2]
```diff
@@ -90,7 +91,16 @@ export class QwcFlywayDatasources extends QwcHotReloadElement {
                 </vaadin-button>`;
         }
     }
-    
+
+    _renderUpdateButton(ds) {
+        if(ds.hasMigrations){
+            return html`
+                <vaadin-button theme="small" @click=${() => this._showCreateDialog(ds)} class="button" title="Create update migration file. Always manually review the created file as it can cause data loss">
```
Comment (yrodiere) [inline extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js:98]: I think you need to add a different dialog, because the one you're opening is designed for the "create" button and will just create the initial migration file, not the update one.

**Review by yrodiere — CHANGES_REQUESTED** [review]
Actually, testing this locally, it seems more work is needed :) See below.

**Comment by cescoffier** [comment]
@rmanibus ping?

**Comment by rmanibus** [comment]
Sorry for the delay, I have been swamped lately. Rebasing now

File: extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js (hunk 2 above)
Comment (rmanibus) [reply extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js]: this is now fixed

File: extensions/agroal/spi/src/main/java/io/quarkus/agroal/spi/JdbcUpdateSQLGeneratorBuildItem.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package io.quarkus.agroal.spi;
+
+import java.util.function.Supplier;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class JdbcUpdateSQLGeneratorBuildItem extends MultiBuildItem {
```
Comment (gastaldi) [inline extensions/agroal/spi/src/main/java/io/quarkus/agroal/spi/JdbcUpdateSQLGeneratorBuildItem.java:12]: Add some javadoc explaining what is this for, so it can be listed in https://quarkus.io/guides/all-builditems

Comment (rmanibus) [reply extensions/agroal/spi/src/main/java/io/quarkus/agroal/spi/JdbcUpdateSQLGeneratorBuildItem.java:12]: @gastaldi done

**Comment by rmanibus** [comment]
@cescoffier @lucamolteni Are we good to merge this ?

**Comment by lucamolteni** [comment]
> @lucamolteni Are we good to merge this ?

can you rebase please?

**Comment by rmanibus** [comment]
> > @lucamolteni Are we good to merge this ?
> 
> can you rebase please?

done !

**Comment by lucamolteni** [comment]
@FroMage ^

**Comment by FroMage** [comment]
Oh, this looks nice. I was expecting a CLI like `quarkus db:update` but this is a nice start :)

**Comment by yrodiere** [comment]
> Oh, this looks nice. I was expecting a CLI like `quarkus db:update` but this is a nice start :)

It probably can be part of the dev mode CLI rather easily. I think the `quarkus` CLI is another story -- there no app or environment when it runs, so it's a bit more complicated.

BTW I wonder how much sense it would make to expose @phillip-kruger 's "MCP tools" as commands one can invoke directly in the dev mode CLI? Then we could hit two birds with one stone, because things like this schema update could in theory make sense for both...

**Comment by phillip-kruger** [comment]
This should be possible if:
- The user's app is running in dev mode and
- Dev MCP is enabled

**Comment by FroMage** [comment]
So you're saying we can't do this from a CLI unless Quarkus is running? OTOH `hibernate-processor` seems to be able to bootstrap at least _some_ of the Hibernate model without a running app, but perhaps not enough for us…

**Comment by yrodiere** [comment]
> So you're saying we can't do this from a CLI unless Quarkus is running? OTOH `hibernate-processor` seems to be able to bootstrap at least _some_ of the Hibernate model without a running app, but perhaps not enough for us…

I would encourage you to have a look at that Hibernate model that hibernate processor is using... It's really bare :)

**Comment by FroMage** [comment]
> I would encourage you to have a look at that Hibernate model that hibernate processor is using... It's really bare :)

Yeah, I figured. Well I guess the CLI can start quarkus if it's not running, or use it if running. Though it'd be nice to be able to start Quarkus "partially" just to get ORM started, that'd be faster, but I doubt it's worth the effort :-/

**Comment by lucamolteni** [comment]
@phillip-kruger from a quick test the feature was working fine, except if I called my flyway script V2.sql I'd get another file called V2-timestamp.sql instead of V3-timestamp.sql, which is not sure is what you wanted to achieve

In general I'd unit test that part to make sure it's working properly

I'm currently busy with an urgent bug so I can take a deep look at it, but can you try to see if you can reproduce the error?

File: extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java [hunk 4]
```diff
@@ -168,6 +176,50 @@ public FlywayActionResponse create(String ds) {
...
+            if (resourcesDir.isEmpty()) {
+                return new FlywayActionResponse("error", "No resource directory found");
+            }
+            // In the current project only
+            Path path = resourcesDir.get(0);
+
+            Path migrationDir = path.resolve(locations.get(0));
+            Files.createDirectories(migrationDir);
+            SimpleDateFormat format = new SimpleDateFormat("yyyy.MM.dd.HHmmss");
+            String timestamp = format.format(new Timestamp(System.currentTimeMillis()));
```
Comment (lucamolteni) [inline extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java]: `new Timestamp(System.currentTimeMillis())`

doesn't handle time zone correctly 

I'd use LocalDateTime instead https://stackoverflow.com/questions/32437550/whats-the-difference-between-instant-and-localdatetime

**Comment by yrodiere** [comment]
> @phillip-kruger from a quick test the feature was working fine, except if I called my flyway script V2.sql I'd get another file called V2-timestamp.sql instead of V3-timestamp.sql, which is not sure is what you wanted to achieve
> 
> In general I'd unit test that part to make sure it's working properly
> 
> I'm currently busy with an urgent bug so I can take a deep look at it, but can you try to see if you can reproduce the error?

I think you meant to ping @rmanibus

**Comment by rmanibus** [comment]
It should be incrementing the minor version not the major. I am surprised there is a dash in there though, it should be a dot

File: extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java (hunk 4 above)
Comment (rmanibus) [reply extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java]: @lucamolteni it's done !

**Review by lucamolteni — CHANGES_REQUESTED** [review]
I think you meant `format` here not `formatter`

**Comment by rmanibus** [comment]
> I think you meant `format` here not `formatter`

ooops, fixed now

**Comment by lucamolteni** [comment]
@rmanibus why did you close this?

**Comment by rmanibus** [comment]
@lucamolteni I am not sure, I did not close it intentionally. 
When I try to reopen, the button is grayed out with a floating error message saying:
"the fix_43723_hibernate_update branch was force pushed or recreated"

**Comment by rmanibus** [comment]
did I rebase on the wrong branch ?

**Comment by rmanibus** [comment]
someone having a similar issue here:
https://irclogs.alpinelinux.org/%23alpine-devel-2019-06.log

**Comment by rmanibus** [comment]
ok it's reopened and rebased

**Review by lucamolteni — APPROVED** [review]
Can you please update this part of the documentation as well https://quarkus.io/guides/hibernate-orm#flyway

To add this feature as well? 

Thank you

**Comment by lucamolteni** [comment]
Also I forgot: 

please run 

```
 mvn -Dquickly clean install -DskipTests=true
```

from the root because these files aren't properly formatted

```
	modified:   extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/devui/FlywayDevUIProcessor.java
	modified:   extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
	modified:   extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/dev/HibernateOrmDevUIProcessor.java
```

**Comment by rmanibus** [comment]
@lucamolteni done, is the doc ok for you ?

**Comment by lucamolteni** [comment]
> @lucamolteni done, is the doc ok for you ?

Text is right but I see failures in the CI, check those. 
While reading it I realised your code will fail with 2026 minor releases :) 

If you have a 

V.1.2026.whatever.sql

file inside your migration, the time stamp will be before the last migration. 

While I can see it's an edge case, I'm still not convinced this time stamp naming is the correct approach. 

Screenshot 2025-09-24 at 17 16 34

As I suggested you in the past 

> In general I'd unit test that part to make sure it's working properly

I still think it's a good idea that you extract the logic of increasing the release inside a single method you can unit test, so that you can easily test these cases

**Comment by lucamolteni** [comment]
@rmanibus hey are you still working on this? :)

**Comment by rmanibus** [comment]
@lucamolteni I’m still working on this.

I understand your point about the naming convention. It’s a tricky topic and has sparked many debates on past projects I’ve worked on.

On larger projects, using integers to index migrations often leads to collisions, since several developers may start their work around the same time and pick the same index. In my experience, using timestamps has been a decent way to avoid those conflicts, as long as the migrations remain independent and can be applied out of order.

Maybe we could even consider allowing users to plug in their own naming strategy?

**Comment by lucamolteni** [comment]
> @lucamolteni I’m still working on this.
> 
> I understand your point about the naming convention. It’s a tricky topic and has sparked many debates on past projects I’ve worked on.
> 
> On larger projects, using integers to index migrations often leads to collisions, since several developers may start their work around the same time and pick the same index. In my experience, using timestamps has been a decent way to avoid those conflicts, as long as the migrations remain independent and can be applied out of order.
> 
> Maybe we could even consider allowing users to plug in their own naming strategy?


Pluggable seems even more complicated. 

Let's go with the timestamps, I approve this

Please provide a unit test of the numbering algorithm

File: extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java [hunk 5]
```diff
@@ -168,6 +176,50 @@ public FlywayActionResponse create(String ds) {
...
+            }
+
+            List<Path> resourcesDir = DevConsoleManager.getHotReplacementContext().getResourcesDir();
+            if (resourcesDir.isEmpty()) {
+                return new FlywayActionResponse("error", "No resource directory found");
+            }
+            // In the current project only
+            Path path = resourcesDir.get(0);
+
+            Path migrationDir = path.resolve(locations.get(0));
```
Comment (lucamolteni) [inline extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java:204]: This part should be unit-tested

**Review by lucamolteni — APPROVED** [review]

**Review by yrodiere — APPROVED** [review]
I rebased and force-pushed. I say let's merge as is :)

**Comment by yrodiere** [comment]
I pushed another commit to fix the warning in the docs, and improve the structure a bit.

**Comment by yrodiere** [comment]
Pushed a change to deal with Windows line endings...
